client/session.go

```diff
@@ -563,3 +595,936 @@ func (s *session) Close() error {
+				currStart[i].blocks[idx].reattempt.attempted =
+					append(currStart[i].blocks[idx].reattempt.attempted, peer)
+				currStart[i].blocks[idx].reattempt.peersMetadata = peersMetadata
+
+				// Leave the block in the current peers blocks list
+				continue
+			}
+
+			// Removing this block
+			blocksLen := len(currStart[i].blocks)
+			idx := currStart[i].idx
+			tailIdx := blocksLen - 1
+			currStart[i].blocks[idx], currStart[i].blocks[tailIdx] = currStart[i].blocks[tailIdx], currStart[i].blocks[idx]
+			currStart[i].blocks = currStart[i].blocks[:blocksLen-1]
+		}
+	}
+
+	return nil
+}
+
+func (s *session) streamBlocksBatchFromPeer(
+	retrier xretry.Retrier,
+	peer hostQueue,
+	batch []*blocksMetadata,
+	blocksResult *blocksResult,
+	enqueueCh *enqueueChannel,
+) {
+	// Prepare request
+	var (
+		req    = rpc.NewFetchBlocksRequest()
+		result *rpc.FetchBlocksResult_
+	)
+	for i := range batch {
+		var starts []int64
+		for j := range batch[i].blocks {
+			starts = append(starts, batch[i].blocks[j].start.UnixNano())
+		}
+		req.Elements = append(req.Elements, &rpc.FetchBlocksParam{
+			ID:     batch[i].id,
+			Starts: starts,
+		})
+	}
+
+	// Attempt request
+	if err := retrier.Attempt(func() error {
+		client, err := peer.GetConnectionPool().NextClient()
+		if err != nil {
+			return err
+		}
+
+		tctx, _ := thrift.NewContext(s.streamBlocksBatchTimeout)
+		result, err = client.FetchBlocks(tctx, req)
+		return err
+	}); err != nil {
+		for i := range batch {
+			s.streamBlocksReattemptFromPeers(batch[i].blocks, enqueueCh)
+		}
+		return
+	}
+
+	// Parse and act on result
+	for i := range result.Elements {
+		if i >= len(batch) {
+			s.log.Errorf("stream blocks response from peer %v returned more IDs than expected", peer.Host().String())
+			break
+		}
+
+		id := result.Elements[i].ID
+		if id != batch[i].id {
+			s.streamBlocksReattemptFromPeers(batch[i].blocks, enqueueCh)
+			s.log.WithFields(
+				xlog.NewLogField("expectedID", batch[i].id),
+				xlog.NewLogField("actualID", id),
+				xlog.NewLogField("indexID", i),
+			).Errorf("stream blocks response from peer %v returned mismatched ID", peer.Host().String())
+			continue
+		}
+
+		for j := range result.Elements[i].Blocks {
+			if j >= len(batch[i].blocks) {
+				s.log.Errorf("stream blocks response from peer %v returned more blocks than expected", peer.Host().String())
+				break
+			}
+
+			block := result.Elements[i].Blocks[j]
+			if block.Start != batch[i].blocks[j].start.UnixNano() {
+				failed := []blockMetadata{batch[i].blocks[j]}
+				s.streamBlocksReattemptFromPeers(failed, enqueueCh)
+				s.log.WithFields(
+					xlog.NewLogField("id", id),
+					xlog.NewLogField("expectedStart", batch[i].blocks[j].start.UnixNano()),
+					xlog.NewLogField("actualStart", block.Start),
+					xlog.NewLogField("indexID", i),
+					xlog.NewLogField("indexBlock", j),
+				).Errorf("stream blocks response from peer %v returned mismatched ID", peer.Host().String())
+				continue
```

Comment:
Is it worth continuing if one of the starts are off? Does that screw up the whole block or do we want to return a partial block?